### PR TITLE
Events: posts, terms

### PR DESCRIPTION
**Pull request by mykytadudariev**

### CODE REVIEW

File: src/classes/vocabulary/jsonld/class-post-jsonld.php [hunk 1]
```diff
@@ -115,7 +154,75 @@ public function wl_after_get_jsonld( $jsonld ) {
...
+		// Fetch the initial 'about' and 'mentions' counts from post meta.
+		$counts = array(
+			'about'    => get_post_meta( $post_id, 'wl_about_count', true ) ? get_post_meta( $post_id, 'wl_about_count', true ) : 0,
+			'mentions' => get_post_meta( $post_id, 'wl_mentions_count', true ) ? get_post_meta( $post_id, 'wl_mentions_count', true ) : 0,
+		);
+
+		// Iterate over the counts array.
+		foreach ( $counts as $key => $count ) {
+			// Check if data has 'about' or 'mentions' and the count is different from the existing meta value.
+			if ( ! empty( $data[ $key ] ) ) {
```
Comment (ziodave) [inline src/classes/vocabulary/jsonld/class-post-jsonld.php:189]: I think we're missing a case:

1. I have x mentions
2. Event has been sent
3. x mentions recorded in post meta
4. Now mentions is not present anymore (i.e. `empty( $data[ $key ] ) === true`)
5. You should send a count of mentions = 0, the current code won't process it

Same happens later for terms

File: src/classes/vocabulary/jsonld/class-term-jsonld.php [hunk 2]
```diff
@@ -39,4 +69,94 @@ public function wl_term_jsonld_array( $jsonld_array, $term_id ) {
...
+		// Fetch the initial 'about' and 'mentions' counts from term meta.
+		$counts = array(
+			'about'    => get_term_meta( $term_id, 'wl_about_count', true ) ? get_term_meta( $term_id, 'wl_about_count', true ) : 0,
+			'mentions' => get_term_meta( $term_id, 'wl_mentions_count', true ) ? get_term_meta( $term_id, 'wl_mentions_count', true ) : 0,
+		);
+
+		// Iterate over the counts array.
+		foreach ( $counts as $key => $count ) {
+			// Check if data has 'about' or 'mentions' and the count is different from the existing meta value.
+			if ( ! empty( $data[ $key ] ) ) {
```
Comment (ziodave) [inline src/classes/vocabulary/jsonld/class-term-jsonld.php:102]: Same as https://github.com/insideout10/wordlift-plugin/pull/1701/files?diff=unified&w=1#diff-bc6fb5d934e6959cc998029202a88d82ec03220e8b5bc8cacd773c34c76ac54bR189

File: src/classes/vocabulary/jsonld/class-post-jsonld.php [hunk 3]
```diff
@@ -115,7 +154,75 @@ public function wl_after_get_jsonld( $jsonld ) {
...
+	 *
+	 * @param $jsonld_arr array The final jsonld before outputting to page.
+	 * @param $post_id int The post id for which the jsonld is generated.
+	 * @param $context int A context for the JSON-LD generation, valid values in Jsonld_Context_Enum
+	 *
+	 * @return array
+	 */
+	public function set_events_request( $jsonld_arr, $post_id, $context ) {
+		// If context is not PAGE or the array is empty, return early.
+		if ( Jsonld_Context_Enum::PAGE !== $context || empty( $jsonld_arr[0] ) ) {
```
Comment (ziodave) [inline src/classes/vocabulary/jsonld/class-post-jsonld.php:170]: If the array is empty, but before we sent a count of mentions and about then we should send an updated count of 0 for both.

File: src/classes/vocabulary/jsonld/class-term-jsonld.php [hunk 4]
```diff
@@ -39,4 +69,94 @@ public function wl_term_jsonld_array( $jsonld_array, $term_id ) {
...
+		if ( null === $id ) {
+			return isset( $_SERVER['REQUEST_URI'] ) ? filter_var( wp_unslash( $_SERVER['REQUEST_URI'] ), FILTER_SANITIZE_URL ) : '';
+		}
+
+		$maybe_url = get_term_meta( $id, Wordlift_Url_Property_Service::META_KEY, true );
+		if ( ! empty( $maybe_url ) ) {
+			return $maybe_url;
+		}
+
+		return get_term_link( $id );
```
Comment (ziodave) [inline src/classes/vocabulary/jsonld/class-term-jsonld.php:159]: We should really get the current page here, I am not sure what's happening here?

**Review by ziodave — CHANGES_REQUESTED** [review]
All the new code that you wrote must be moved to a module in src/modules/events

**Review by ziodave — CHANGES_REQUESTED** [review]

File: src/modules/events/includes/Options_Entity/Events_Options_Entity_Include_Exclude.php [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	 *
+	 * @access private
+	 * @var Api_Service
+	 */
+	private $api_service;
+
+	/**
+	 * @param Api_Service $api_service
+	 */
+	public function __construct( Api_Service $api_service ) {
```
[the remark this answers is not in the document]
Comment (mykytadudariev) [reply src/modules/events/includes/Options_Entity/Events_Options_Entity_Include_Exclude.php:25]: Sure, no problem, I just copied that "logic" from an existing module [Food Kg](https://github.com/insideout10/wordlift-plugin/blob/main/src/modules/food-kg/includes/Main_Entity/Food_Kg_Main_Entity_Runner.php#L28). I mean the same "behavior"

File: src/modules/events/includes/Options_Entity/Events_Options_Entity_Include_Exclude.php [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+	/**
+	 * Include exclude event update.
+	 *
+	 * @throws Exception If the application fails to load the services configuration file or if the URL cannot be processed.
+	 */
+	public function event_update() {
+		// Get the configurations.
+		$config     = get_option( 'wl_exclude_include_urls_settings', array() );
+		$old_config = get_option( 'wl_exclude_include_urls_settings_old', array() );
```
[the remark this answers is not in the document]
Comment (mykytadudariev) [reply src/modules/events/includes/Options_Entity/Events_Options_Entity_Include_Exclude.php:45]: ok, no problem, I did this logic due to your requirements with `add_action( 'update_option_wl_exclude_include_urls_settings', ... )` like in the document. But no problem, let's rebuild that with another hook

**Pull request merged**